### ISSUE TITLE
build.sh fails with Visual Studio 2017

### ISSUE THREAD

**Issue by moodmosaic** [hedgehogqa/fsharp-hedgehog#108]
I setup a Virtual Machine with Windows 10, Visual Studio 2017 and making sure to select F# in the installer.

Even though I can compile the solution in Visual Studio 2017, when I run `./build.sh` in Git Bash it fails:

```
$ ./build.sh
Paket version 5.1.8
Performance:
 - Disk IO: 92 milliseconds
 - Runtime: 1 second
Building project with version: LocalBuild
Shortened DependencyGraph for Target Test:
<== Test
   <== Doctest
      <== Build

The resulting target order is:
 - Build
 - Doctest
 - Test
Starting Target: Build
Building project: C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln
  c:\Windows\Microsoft.NET\Framework\v4.0.30319\MSBuild.exe  C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln /t:Rebuild /m      /p:RestorePackages="False" /p:Configuration="Release" /logger:Fake.MsBuildLogger+ErrorLogger,"C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\packages\FAKE\tools\FakeLib.dll"
Microsoft (R) Build Engine version 4.7.2046.0
[Microsoft .NET Framework, version 4.0.30319.42000]
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 6/23/2017 2:16:16 PM.
     1>Project "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln" on node 1 (Rebuild target(s)).
     1>ValidateSolutionConfiguration:
         Building solution configuration "Release|Any CPU".
     1>Project "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln" (1) is building "C:\Users\Nikos\Documen
       ts\Hedgehog\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj" (2) on node 1 (Rebuild target(s)).
     2>C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj(42,3): error MSB4019: The importe
       d project "C:\Program Files (x86)\Microsoft SDKs\F#\3.0\Framework\v4.0\Microsoft.FSharp.Targets" was not found.
       Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
     2>Done Building Project "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj" (Rebuild
       target(s)) -- FAILED.
     1>Project "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln" (1) is building "C:\Users\Nikos\Documen
       ts\Hedgehog\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" (3) on node 2 (Rebuild target(s)).
     3>C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj(51,3): error MSB410
       2: The value "" of the "Project" attribute in element <Import> is invalid. Parameter "path" cannot have zero len
       gth.
     3>Done Building Project "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsp
       roj" (Rebuild target(s)) -- FAILED.
     1>Done Building Project "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln" (Rebuild target(s)) -- FA
       ILED.

Build FAILED.

       "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln" (Rebuild target) (1) ->
       "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj" (Rebuild target) (2) ->
         C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\src\Hedgehog\Hedgehog.fsproj(42,3): error MSB4019: The impor
       ted project "C:\Program Files (x86)\Microsoft SDKs\F#\3.0\Framework\v4.0\Microsoft.FSharp.Targets" was not found
       . Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.


       "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln" (Rebuild target) (1) ->
       "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj" (Rebuild target)
       (3) ->
         C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\tests\Hedgehog.Tests\Hedgehog.Tests.fsproj(51,3): error MSB4
       102: The value "" of the "Project" attribute in element <Import> is invalid. Parameter "path" cannot have zero l
       ength.

    0 Warning(s)
    2 Error(s)

Time Elapsed 00:00:00.76
Running build failed.
Error:
Building C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Hedgehog.sln failed with exitcode 1.

---------------------------------------------------------------------
Build Time Report
---------------------------------------------------------------------
No target was successfully completed
---------------------------------------------------------------------
```

**Comment by moodmosaic** [comment]
It looks like Visual Studio 2017 installs Visual F# 4.1, so all I needed was to install [Visual F# 4.0](https://www.microsoft.com/en-us/download/confirmation.aspx?id=48179) as well.

---

```
Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:05.95
Finished Target: Build
Starting Target: Doctest (==> Build)
C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\Doctest already exists.
Finished Target: Doctest
Starting Target: Test (==> Doctest)
C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\packages\xunit.runner.console\tools\xunit.console.exe "C:\Users\Nikos\Documents\Hedgehog\fsharp-hedgehog\tests\Hedgehog.Tests\bin\Release\Hedgehog.Tests.dll" -parallel all
xUnit.net Console Runner (64-bit .NET 4.0.30319.42000)
  Discovering: Hedgehog.Tests
  Discovered:  Hedgehog.Tests
  Starting:    Hedgehog.Tests
  Finished:    Hedgehog.Tests
=== TEST EXECUTION SUMMARY ===
   Hedgehog.Tests  Total: 135, Errors: 0, Failed: 0, Skipped: 0, Time: 37.512s
Finished Target: Test

---------------------------------------------------------------------
Build Time Report
---------------------------------------------------------------------
Target     Duration
------     --------
Build      00:00:06.2639642
Doctest    00:00:05.3679847
Test       00:00:41.5872624
Total:     00:00:53.2891169
Status:    Ok
---------------------------------------------------------------------
```

**Comment by moodmosaic** [comment]
I think, we have to modify our *.fsproj file according to [this](https://github.com/Microsoft/visualfsharp/issues/2340).

**Comment by ploeh** [comment]
I can reproduce the problem using Visual Studio 2017 Community Edition. Digging around a bit, I found that my machine has an `Microsoft.FSharp.Targets` file in `C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\Microsoft\VisualStudio\v15.0\FSharp`.

The entire contents of that file, however, is:

```
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  
  <Import Project="$(MSBuildProgramFiles32)\Microsoft SDKs\F#\4.1\Framework\v4.0\Microsoft.FSharp.Targets" />
  
</Project>
```

In other words, it looks like I have two options for including the file:

Either

```
<Import Project="C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\Microsoft\VisualStudio\v15.0\FSharp\Microsoft.FSharp.Targets" />
```

or 

```
<Import Project="$(MSBuildProgramFiles32)\Microsoft SDKs\F#\4.1\Framework\v4.0\Microsoft.FSharp.Targets" />
```

Both options seem to work, in the sense that `build.sh` then completes successfully.

From @dsyme's [comment](https://github.com/Microsoft/visualfsharp/issues/2340#issuecomment-298169222)

> My impression is that this is being dealt with via updates to FAKE and by using the right msbuild on a system.

I get the impression that this ought to 'just work'. Clearly, it doesn't, but at this point, I'm not even sure I'm looking in the right direction. @dsyme, can you advise, given that you closed https://github.com/Microsoft/visualfsharp/issues/2340 ?

**Comment by dsyme** [comment]
@moodmosaic @ploeh I';m not totally sure about building from bash.  But first update your FAKE version and try again?

**Comment by ploeh** [comment]
@dsyme, thank you for responding so quickly. The issue's not related to Bash. It fails in the same way if I try to run `build.cmd` from a standard command prompt.

If by "update your FAKE version" you mean

```
.paket\paket.exe update
```

then that makes no differences.

**Comment by moodmosaic** [comment]
@dsyme, @ploeh, FTR, I just downloaded the installer for Visual Studio 2017 Community, and installed it side-by-side on a machine which has already Visual Studio 2015 Community (Update 3) installed.

In Visual Studio 2017 Community installer, I selected F#, which auto-selects .NET 4.6.x (among others). Hedgehog's target framework version is .NET 4.5.2, which wasn't auto-selected so I've manually selected it.

Then, I pulled the latest from `master` branch and did `git clean -xdf` followed by a `build.sh` which appears to complete successfully (!) 😕 

I do believe that the issue exists though, so it's very helpful that @ploeh is able to reproduce it.

**Comment by ploeh** [comment]
Hmm...

![image](https://user-images.githubusercontent.com/242165/31492233-7238a248-af4a-11e7-9178-7a927583d180.png)

**Comment by moodmosaic** [comment]
Perhaps... If you choose to install more items, other than the F#.

I’ve only clicked on the F# checkbox, and that caused .NET 4.6.1 (if I’m not mistaken) to be checked as well. (It also caused C# and other items to be checked, but that should be fine.)

Then, I clicked on the .NET 4.5.2 checkbox and then I’ve started the installation.

**Comment by moodmosaic** [comment]
I’m guessing that there’s some conflict caused by some of those selected components.

I’ve also had the same issue in the past (because I’ve created this issue here on GitHub), so there’s definitely something wrong, ...somewhere.

**Comment by dsyme** [comment]
@ploeh does build.cmd from a VS developer command prompt work?

**Comment by ploeh** [comment]
@dsyme, no, it produces the same error message.

FWIW, the machine in question is one I wiped and rebuilt from scratch last weekend. I don't expect that there's a lot of cruft on it; rather, I suspect that it may be too pristine.

For instance, as @moodmosaic has pointed out earlier in this thread, it looks as though one can fix the problem by installing F# 4.0. I'd have no problem taking an action like that, but since I find myself in that (for me) rare situation of having a 'fresh' machine, I thought it'd be worthwhile to report the problem.

What I mean is that if I'm having a problem with a freshly installed machine, could others experience the same problem as well?

**Comment by Porges** [comment]
FWIW I just tried running build.cmd and since the output said:

    "C:\Program Files (x86)\Microsoft SDKs\F#\3.0\Framework\v4.0\Microsoft.FSharp.Targets"
    was not found.

I installed F# 3.0. This still failed to build (I think it misinterpreted some syntax), so I've uninstalled that and installed 4.0. Now I get a different error:

    "C:\Program Files (x86)\MSBuild\Microsoft\VisualStudio\v12.0\FSharp\Microsoft.FSharp.Targets"
    was not found.

I may have broken my setup? :grin: Everything works just fine from inside VS2017...

**Comment by Porges** [comment]
So, updating FAKE to the latest stable version made it a bit better, it tried to use the 2017 targets;

    "C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\Microsoft\VisualStudio\v\FSharp\Microsoft.FSharp.Targets"
    was not found.

However, the actual version number is missing from the path...

**Comment by moodmosaic** [comment]
I'm closing this, since we've now merged #153—released as Hedgehog 0.7. Please reopen this if necessary, or feel free to add comments to it.
